src/main/java/seedu/address/logic/commands/DeleteCommand.java

```diff
@@ -19,30 +31,93 @@ public class DeleteCommand extends Command {
     }
 
     @Override
     public CommandResult execute(Model model) throws CommandException {
         requireNonNull(model);
         List<Person> lastShownList = model.getFilteredPersonList();
 
-        if (targetIndex.getZeroBased() >= lastShownList.size()) {
-            throw new CommandException(Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX);
+        Optional<Person> findPersonToDelete = null;
+
+        if (name != null) {
+            Stream<Person> personbyName = lastShownList.stream()
+                    .filter(person -> person.getName().equals(name));
+            findPersonToDelete = findPersonToDeleteName(personbyName);
+        } else if (nric != null) {
+            findPersonToDelete = findPersonToDeleteIc(lastShownList);
+        }
+
+        Person personToDelete = findPersonToDelete.get();
+
+        if (deletePersonDescriptor.isAllFalse()) {
+            model.deletePerson(personToDelete);
+            return new CommandResult(String.format(MESSAGE_DELETE_PERSON_SUCCESS, Messages.format(personToDelete)));
+        } else {
+            Person editedPerson = createEditedPerson(personToDelete, deletePersonDescriptor);
+            model.setPerson(personToDelete, editedPerson);
+            model.updateFilteredPersonList(PREDICATE_SHOW_ALL_PERSONS);
+            return new CommandResult(String.format(MESSAGE_DELETE_PERSON_FIELD_SUCCESS, Messages.format(editedPerson)));
+        }
+    }
+
+    /**
+     * Finds the person to delete by NRIC.
+     *
+     * @param persons list of persons in the address book.
+     * @return the person to delete.
+     * @throws CommandException if the person is not found.
+     */
+    public Optional<Person> findPersonToDeleteIc(List<Person> persons)
+            throws CommandException {
+        Optional<Person> personbyNric = persons.stream()
+                .filter(person -> person.getNric().equals(nric))
+                .findFirst();
+        if (personbyNric.isPresent()) {
+            return personbyNric;
+        } else {
+            throw new CommandException(Messages.MESSAGE_INVALID_NRIC);
         }
+    }
 
-        Person personToDelete = lastShownList.get(targetIndex.getZeroBased());
-        model.deletePerson(personToDelete);
-        return new CommandResult(String.format(MESSAGE_DELETE_PERSON_SUCCESS, Messages.format(personToDelete)));
+    /**
+     * Finds the person to delete by name.
+     *
+     * @param persons list of persons in the address book.
+     * @return the person to delete.
+     * @throws CommandException if the person is not found.
+     */
+    public Optional<Person> findPersonToDeleteName(Stream<Person> persons)
+            throws CommandException {
+        Optional<Person> personbyName = persons.findFirst();
+        if (personbyName.isPresent()) {
+            return personbyName;
+        } else {
+            throw new CommandException(Messages.MESSAGE_INVALID_NAME);
+        }
```

Comment:
See [this comment](https://github.com/AY2324S1-CS2103T-T14-3/tp/pull/77/files#r1363750031) I left for `EditCommand` - to summarize, this isn't an urgent issue, but more so an note - it might be better to use an interface for the utility method, but we can leave it until after the varying features are merged.